File: src/ai_multi_reference_timekeeping/safety.py
"""Safety case artifacts inspired by military and aerospace standards.

This module provides lightweight hazard tracking and risk scoring aligned with
MIL-STD-882E, DO-178C, NASA NPR 7150.2D, MIL-STD-498, STANAG 4626, IEC 61508,
and NASA Power of 10 principles. It is intentionally simple and explicit to
remain analyzable and auditable.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Iterable


@dataclass(frozen=True)
class Hazard:
    """Hazard definition with severity and likelihood.

    Severity and likelihood follow MIL-STD-882E qualitative levels:
    - severity: 1 (catastrophic) .. 4 (negligible)
    - likelihood: A (frequent) .. E (improbable)
    """

    code: str
    description: str
    severity: int
    likelihood: str
    mitigation: str


@dataclass
class HazardOccurrence:
    """Record of a hazard occurrence."""

    code: str
    detail: str
    timestamp: float


class RiskMatrix:
    """Simple MIL-STD-882E risk matrix."""

    _order = {"A": 5, "B": 4, "C": 3, "D": 2, "E": 1}

    @classmethod
    def score(cls, severity: int, likelihood: str) -> int:
        # Higher score indicates higher risk.
        if severity < 1 or severity > 4:
            raise ValueError("severity must be 1..4")
        if likelihood not in cls._order:
            raise ValueError("likelihood must be A..E")
        return (5 - severity) * cls._order[likelihood]

    @classmethod
    def acceptable(cls, severity: int, likelihood: str, threshold: int = 6) -> bool:
        # Acceptable risk threshold is configurable.
        return cls.score(severity, likelihood) <= threshold


@dataclass
class SafetyCase:
    """Safety case state with hazard registry and occurrence log."""

    hazards: dict[str, Hazard] = field(default_factory=dict)
    occurrences: list[HazardOccurrence] = field(default_factory=list)

    def register(self, hazard: Hazard) -> None:
        # Register hazards before recording occurrences.
        self.hazards[hazard.code] = hazard

    def record(self, code: str, detail: str, timestamp: float) -> None:
        # Record hazard occurrence with timestamp and detail.
        if code not in self.hazards:
            raise ValueError(f"Unknown hazard code {code}")
        self.occurrences.append(HazardOccurrence(code=code, detail=detail, timestamp=timestamp))

    def risk_summary(self) -> dict[str, int]:
        # Provide a risk score per hazard for reporting.
        return {
            code: RiskMatrix.score(hazard.severity, hazard.likelihood)
            for code, hazard in self.hazards.items()
        }

    def unacceptable_hazards(self, threshold: int = 6) -> list[Hazard]:
        # Filter hazards that exceed acceptable risk.
        return [
            hazard
            for hazard in self.hazards.values()
            if not RiskMatrix.acceptable(hazard.severity, hazard.likelihood, threshold=threshold)
        ]

File: cpp/src/safety.cpp
#include "aimrt/safety.hpp"

#include <stdexcept>

namespace aimrt {

namespace {

int likelihood_score(const std::string& likelihood) {
    if (likelihood == "A") {
        return 5;
    }
    if (likelihood == "B") {
        return 4;
    }
    if (likelihood == "C") {
        return 3;
    }
    if (likelihood == "D") {
        return 2;
    }
    if (likelihood == "E") {
        return 1;
    }
    throw std::runtime_error("likelihood must be A..E");
}

}  // namespace

int RiskMatrix::score(int severity, const std::string& likelihood) {
    if (severity < 1 || severity > 4) {
        throw std::runtime_error("severity must be 1..4");
    }
    return (5 - severity) * likelihood_score(likelihood);
}

bool RiskMatrix::acceptable(int severity, const std::string& likelihood, int threshold) {
    return score(severity, likelihood) <= threshold;
}

void SafetyCase::register_hazard(const Hazard& hazard) {
    hazards_[hazard.code] = hazard;
}

void SafetyCase::record(const std::string& code, const std::string& detail, double timestamp) {
    if (hazards_.find(code) == hazards_.end()) {
        throw std::runtime_error("Unknown hazard code " + code);
    }
    occurrences_.push_back(HazardOccurrence{code, detail, timestamp});
}

std::map<std::string, int> SafetyCase::risk_summary() const {
    std::map<std::string, int> summary;
    for (const auto& [code, hazard] : hazards_) {
        summary[code] = RiskMatrix::score(hazard.severity, hazard.likelihood);
    }
    return summary;
}

std::vector<Hazard> SafetyCase::unacceptable_hazards(int threshold) const {
    std::vector<Hazard> output;
    for (const auto& [code, hazard] : hazards_) {
        if (!RiskMatrix::acceptable(hazard.severity, hazard.likelihood, threshold)) {
            output.push_back(hazard);
        }
    }
    return output;
}

}  // namespace aimrt
